internal/proxy/gsettings.go

```diff
@@ -0,0 +1,197 @@
+			if password, ok := p.url.User.Password(); ok {
+				settings += fmt.Sprintf("authentication-password='%s'\n", escapeSingleQuote(password))
+			}
+		}
+	case protocolNo:
+		// Ignored hosts are configured at the root level
+		section = fmt.Sprintf("[%s]", systemProxySchemaID)
+
+		hosts := strings.Split(p.escapedURL, ",")
+		for i, host := range hosts {
+			hosts[i] = wrapHostIfNeeded(host)
+		}
+		settings = fmt.Sprintf("ignore-hosts=[%s]\n", strings.Join(hosts, ","))
+	case protocolAuto:
+		// Autoconfig URL is configured at the root level
+		section = fmt.Sprintf("[%s]", systemProxySchemaID)
+		settings = fmt.Sprintf("autoconfig-url='%s'\n", p.escapedURL)
+	}
+
+	return fmt.Sprintf("%s\n%s\n", section, settings)
+}
+
+// applyToGSettings applies the proxy configuration in the form of a GSchema override file,
+// then runs glib-compile-schemas to make the changes visible to GSettings.
+// If there are no proxy settings to apply, the GSchema override file is removed.
+func (p Proxy) applyToGSettings() (err error) {
+	defer decorate.OnError(&err, "couldn't apply GSettings proxy configuration")
+
+	// On the off chance that the user is not running GNOME, we want to print a warning and quietly return.
+	if _, err := exec.LookPath(p.glibCompileSchemasCmd[0]); err != nil {
+		log.Warningf("Couldn't find an executable for %q, not applying GSettings proxy configuration", p.glibCompileSchemasCmd[0])
+		return nil
+	}
+
+	// Check if the parent directory exists - fail if it doesn't, as it means we
+	// don't have any defined proxy XML schema to override.
+	if stat, err := os.Stat(filepath.Dir(p.gsettingsConfigPath)); err != nil {
+		return fmt.Errorf("couldn't find GLib schema directory: %w", err)
+	} else if !stat.IsDir() {
+		return fmt.Errorf("GLib schema path %q is not a directory", filepath.Dir(p.gsettingsConfigPath))
+	}
+
+	if len(p.settings) == 0 {
+		log.Debug("No proxy settings to apply, removing GSchema override file if it exists")
+
+		// If we managed to remove something, we need to recompile the schemas
+		// to propagate the change to GSettings.
+		if err := os.Remove(p.gsettingsConfigPath); err == nil {
+			log.Debugf("Removed GSettings override file at %q", p.gsettingsConfigPath)
+			return p.runGlibCompileSchemas()
+		} else if !errors.Is(err, fs.ErrNotExist) {
+			return err
+		}
+		return nil
+	}
+	log.Debugf("Applying GSettings proxy configuration to %q", p.gsettingsConfigPath)
+
+	content := p.gsettingsConfig()
+	prevContent, err := previousConfig(p.gsettingsConfigPath)
+	if err == nil && prevContent == content {
+		log.Debugf("GSettings proxy configuration at %q is already up to date", p.gsettingsConfigPath)
+		return nil
+	} else if err != nil && !errors.Is(err, fs.ErrNotExist) {
+		return err
+	}
+
+	backupPath, moveBack, err := backupExistingFile(p.gsettingsConfigPath)
+	if err != nil {
+		return err
+	}
+
+	if err := safeWriteFile(p.gsettingsConfigPath, content); err != nil {
+		return errors.Join(err, moveBack())
```

Comment:
The error you are wrapping is the error of `moveBack()`, not the one of `safeWriteFile()` or `runGlibCompileSchemas()`. Also, I find it weird that errors.Join() has a side-effect for restoring.

I would just have them in 3 lines:
- `errors.Join()` with the current error
- `moveBack()`
- if error on moveBack, `errors.Join()` it

Reply:
I will need to get used to join a nil error, but reading the standard library, that works :)